File: r-analytics/predict.R
# predict.R
# Loads the trained R model and predicts occupancy 30 minutes ahead for a
# single zone snapshot. R equivalent of ml/predict.py.
#
# Run (from the project root):
#   Rscript r-analytics/predict.R

suppressMessages(library(randomForest))

MODEL_PATH <- "r-analytics/model.rds"
FEATURES <- c(
  "hour", "day_of_week", "is_weekend", "is_holiday",
  "temperature", "rainfall", "current_occupancy", "capacity", "average_duration"
)

risk_level <- function(predicted, capacity) {
  pct <- predicted / capacity * 100
  if (pct >= 90) "HIGH" else if (pct >= 70) "MEDIUM" else "LOW"
}

if (!file.exists(MODEL_PATH)) {
  stop("r-analytics/model.rds not found. Run r-analytics/train_model.R first.")
}

model <- readRDS(MODEL_PATH)

# Kept in-distribution with the demo dataset's zones (capacity 30-150) --
# see the note in ml/predict.py about not extrapolating outside training ranges.
example <- data.frame(
  hour = 18, day_of_week = 4, is_weekend = 0, is_holiday = 0,
  temperature = 29.0, rainfall = 0.0, current_occupancy = 34,
  capacity = 40, average_duration = 65.0
)

predicted <- as.numeric(predict(model, newdata = example))
current_pct <- example$current_occupancy / example$capacity * 100

cat("Input snapshot:\n")
print(example)
cat(sprintf("\nCurrent occupancy  : %.1f%%\n", current_pct))
cat(sprintf(
  "Predicted (30 min) : %.2f vehicles (%.1f%%)\n",
  predicted, predicted / example$capacity * 100
))
cat(sprintf("Risk level         : %s\n", risk_level(predicted, example$capacity)))

File: ml/predict.py
"""
predict.py
----------
Loads the trained model and predicts occupancy 30 minutes ahead for a
single zone snapshot. Useful for a quick demo or for wiring into an API.

Run:
    python ml/predict.py
"""

from pathlib import Path

import joblib
import pandas as pd

MODEL_PATH = Path(__file__).resolve().parent / "model.pkl"

FEATURES = [
    "hour",
    "day_of_week",
    "is_weekend",
    "is_holiday",
    "temperature",
    "rainfall",
    "current_occupancy",
    "capacity",
    "average_duration",
]


def risk_level(predicted, capacity):
    pct = (predicted / capacity) * 100 if capacity else 0
    if pct >= 90:
        return "HIGH"
    if pct >= 70:
        return "MEDIUM"
    return "LOW"


def predict_one(model, sample: dict) -> dict:
    row = pd.DataFrame([sample])[FEATURES]
    predicted = float(model.predict(row)[0])
    return {
        "predicted_occupancy": round(predicted, 2),
        "risk_level": risk_level(predicted, sample["capacity"]),
    }


if __name__ == "__main__":
    if not MODEL_PATH.exists():
        raise SystemExit("ml/model.pkl not found. Run ml/train_model.py first.")

    model = joblib.load(MODEL_PATH)

    # NOTE: keep inputs within the range of your training data (zone capacity,
    # occupancy, durations etc). A Random Forest cannot extrapolate sensibly
    # outside the ranges it was trained on -- if your real zones are bigger
    # than the ones in your training set, retrain on data that covers them.
    example = {
        "hour": 18,
        "day_of_week": 4,
        "is_weekend": 0,
        "is_holiday": 0,
        "temperature": 29.0,
        "rainfall": 0.0,
        "current_occupancy": 34,
        "capacity": 40,
        "average_duration": 65.0,
    }

    result = predict_one(model, example)
    current_pct = example["current_occupancy"] / example["capacity"] * 100

    print("Input snapshot:", example)
    print(f"\nCurrent occupancy      : {current_pct:.1f}%")
    print(f"Predicted (30 min)     : {result['predicted_occupancy']} vehicles "
          f"({result['predicted_occupancy'] / example['capacity'] * 100:.1f}%)")
    print(f"Risk level             : {result['risk_level']}")
